[ML Data Frame] Refactor stop logic

davidkyle opened the pull request:
In `AsyncTwoPhaseIndexer. finishAndSetState` the `onStop` and `onAbort` methods could be called from inside an atomic update. Both those methods are abstract designed to be overridden and that overriding, by an implementor who is not aware of the restrictions (i.e. me), may introduce side effects which are not safe. 

I'm not sure what the behaviour is if another thread tries to `get()` the atomic reference during a call to `updateAndGet()` but the docs warn against it. I believe this is the cause of the CI failures seen in #42344. 

> The function should be side-effect-free, since it may be re-applied when attempted updates fail due to contention among threads.

https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/atomic/AtomicReference.html#updateAndGet(java.util.function.UnaryOperator)

The methods `onStop` and `onFinish` must be called before state is saved (1 to set the correct Data Frame state and 2 to increment the checkpoint) but this ordering means there is a race between `onStop` completing the persistent task and `doSaveState` updating the persistent task parameters. Luckily after #41942 it is no longer necessary to update the persistent task as all state is persisted and restored from the index so I have remove the p. task update from `doSaveState`.

`ClientDataFrameIndexer.onStop` was also persisting state which is not necessary as `doSaveState` is called after `onStop` by the base class.

There is also a refactoring change to `AsyncTwoPhaseIndexer.stop` to make it work the same as `abort`. `abort` does not call `onAbort` if the indexer is not running it is left to the client code to handle that, `stop` should use the same pattern as it is confusing to use 2 different paradigms in the class interface. 

Finally this un-mutes tests muted for #42641 

Closes #42344

davidkyle committed:
Revert "invalid test"

This reverts commit 9dd8b52c13c716918ff97e6527aaf43aefc4695d.

davidkyle committed:
Testing

davidkyle committed:
mend

davidkyle committed:
Revert "[ML Data Frame] Mute Data Frame tests"

This reverts commit 5d837fa312b0e41a77a65462667a2d92d1114567.

davidkyle committed:
Call onStop and onAbort outside atomic update

davidkyle committed:
Don’t update CS

davidkyle committed:
Tidying up

elasticmachine commented:
Pinging @elastic/ml-core

benwtrent reviewed (approved):
Nice catch on the side-effect free functions. 

I also think it is good to keep `stop` and `abort` in behavior parity. But, are we moving in the right direction? Should we move `abort` to call `onAbort` directly, instead of changing `stop`? I am not 100% which is better, but as long as the implementer of the class is aware that, then I suppose it is ok. 

Tests will have to be updated in `AsyncTwoPhaseIndexerTests` to account for `onStop()` not being called when the Indexer is not running. 

We should definitely get a review from @jimczi  or @polyfractal

polyfractal commented:
Sorry, I've not been following recent code changes, still catching up.  Some basic questions below, I may be way off base, happy to be set straight :)

> There is also a refactoring change to AsyncTwoPhaseIndexer.stop to make it work the same as abort. abort does not call onAbort if the indexer is not running it is left to the client code to handle that, stop should use the same pattern as it is confusing to use 2 different paradigms in the class interface.

I'm a little confused.  The model is that someone calls `abort()`, the AsyncIndexer sets internal state to `ABORTING` and waits for the indexer to notice the state has changed.  When that happens, `onAbort()` is called and the indexer implementation tears itself down (deallocating resources, killing the task, etc)

This is in spirit the same as `stop()`: a call to stop just sets the internal state to `STOPPING`/`STOPPED` and waits for the indexer implementation to notice.

Can we back up and revisit what `onStop()` is for and why it needs to be in `finishAndSetState()`?  Why can't these actions take place when `stop()` is called?

> The methods onStop and onFinish must be called before state is saved 

I'm a bit worried about this too, since `onFinish()` is only supposed to be called by the indexer when it reaches the "end of the data".  So there really shouldn't be any dependency on `onStop` and `onFinish`?

>  Both those methods are abstract designed to be overridden and that overriding, by an implementor who is not aware of the restrictions (i.e. me), may introduce side effects which are not safe.

Yeah this was probably a bad design choice on my part, we should probably consider removing `onAbort()` from the atomic update no matter the outcome of the ^^^^ above questions. :)

davidkyle commented:
run elasticsearch-ci/2

davidkyle commented:
>  the AsyncIndexer sets internal state to ABORTING and waits for the indexer to notice the state has changed

Data Frames and Rollups have diverge a little, in particular once a Data Frame has finished it increments a checkpoint and will not be triggered again. 

https://github.com/elastic/elasticsearch/blob/0db0e1330c55b6c5a4c529d6a4f6ecdca7dc4449/x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java#L249

After this point the indexer will not fire again and notice the state has change and trigger the abort. The Java doc says as much and there is code that checks the return value of `abort()` to handle this case. 

https://github.com/elastic/elasticsearch/blob/7c503ce20d762b82da975764482cc6eace542fa2/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexing/AsyncTwoPhaseIndexer.java#L121

> Can we back up and revisit what onStop() is for and why it needs to be in finishAndSetState()? Why can't these actions take place when stop() is called?

We made a change to complete the data frame persistent task when it is stopped #41752. In that I introduced an `onStop` event so the data frame task would know when the indexer has done its work and it can be completed (`stop` is called, some time later the indexer stops then `onStop` is triggered). This followed the pattern of `onAbort` and suffers from the same problem `abort` does in that it is not triggered if the indexer has finished. So I put the fix in the base class then realised this also affects roll ups and changed the code as I described above. 

The difference between `onStop` and `onFinish` it that the first is only called after `stop()` is executed. `onFinish` represents the case where the indexer came to a natural end i.e. it finished its work. There isn't a dependency between `onStop` and `onFinish` if stop is called before the indexer finishes then `onFinish` will not be called. 

>> The methods onStop and onFinish must be called before state is saved

> I'm a bit worried about this too, since onFinish() is only supposed to be called by the indexer when it reaches the "end of the data". So there really shouldn't be any dependency on onStop and onFinish?

What i meant is that the Data Frame task modifies its state on onStop and onFinish so there is a contract that they will be called before `doSaveState` rather than after it (when appropriate) so the changes are captured. 

> we should probably consider removing onAbort()

That's in this PR

davidkyle committed:
Remove invalid test that asserted logic that has been removed

davidkyle committed:
Add stopped event

davidkyle commented:
The plot thickens and this change just got a little more complicated sorry.

When `stop()` is used to stop a data frame 3 things should happen:

1. The data frame should be notified the indexer is stopping prior to persisting state so it can make the relevant changes. This is now the `onStopping()` method
2. The indexer persists the updated state 
3. The rest API call must wait for the the indexer to stop and persist its state so that when it returns any subsequent rest calls will see that the indexer state is `STOPPED. This required a new method `onStopped` which is called after the indexer has stopped and persisted its state.

Initially 1 and 3 where happening in the `onStop` method creating a race between persisting state and returning from the rest call, with the consequence that subsequent calls to check the state may or may not see the last update. Obviously this broke CI and the UI needs to reliably know that stop means stopped.

benwtrent committed:
Revert "Add stopped event"

This reverts commit 02ba992f4818bebd838e1c7678bd2e1cc090bfab.

polyfractal commented:
Thanks for catching me up @davidkyle , appreciate it :)

I don't know enough about dataframes to suggest something useful, but I do worry about these chained state updates and their dependencies.  E.g. if something goes wrong it seems like it could really mess with the "state machine" managing the task and lead to a zombie.  

Rollup (and the async indexer) were mostly designed around "optimistic" checkpoints.  Checkpoint when you can, but if something goes wrong you can just load up from last checkpoint and overwrite a portion of already-written documents with the same data.  That keeps persistence mostly uncoupled from everything else, more of a fire-and-forget background operation.

I know Dataframes is different, but I wonder if perhaps there's a way to do similar so we don't need this chain of dependencies?  Not trying to block the PR, I'm just worried about growing complexity of an already tricky async task thing :)

> After this point the indexer will not fire again and notice the state has change and trigger the abort. 

Could dataframe just perform a "maintenance" check here when triggered, even if there isn't work to do?

benwtrent committed:
Adding check for STOPPED in saveState

davidkyle committed:
Merge branch 'master' into refactor-stop-logic

polyfractal commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexing/AsyncTwoPhaseIndexer.java:
```diff
@@ -315,6 +310,14 @@ private IndexerState finishAndSetState() {
             }
         });
+
+        if (callOnStop.get()) {
```
Since this doesn't need the intermediate `stopping` state anymore, could we just check `updatedState` and avoid the atomic booleans?

I suppose it'd also trigger if was already stopped and that's not what you want?

benwtrent replied:
@polyfractal `stopping` still provides a call to `doSaveState` when `checkState` is called against it. 
But, changing these branches to
```
if(updatedState == IndexerState.STOPPED) {
   onStop();
else if(udpatedState == IndexerState.ABORTING){
   onAbort();
}
```
Maybe could work.

benwtrent replied:
@davidkyle what do you think? That would make it out of step with `onAbort` but, we could possibly remove at least that atomic boolean

polyfractal replied:
Eh, if it requires jumping through hoops it's probably not worth it.  Just was looking to see if there was an easy way to avoid the atomics.

polyfractal replied:
Carry on :+1:

polyfractal reviewed (approved):
Left a comment/question, but the changes to the indexer bit are fine with me.  I didn't look closely at the actual dataframe logic but I assume you know what you're doing there :D

Thanks for helping to simplify things, I appreciate it @davidkyle @benwtrent :heart:
